gitman/shell.py

```diff
@@ -118,11 +118,50 @@ def pwd(_show=True):
     return cwd
 
 
-def ln(source, target):
-    dirpath = os.path.dirname(target)
-    if not os.path.isdir(dirpath):
-        mkdir(dirpath)
-    os.symlink(source, target)
+def ln(source, target, symbolic):
+    if symbolic:
+        dirpath = os.path.dirname(target)
+        if not os.path.isdir(dirpath):
+            mkdir(dirpath)
+        os.symlink(source, target)
+    else:
+        if not os.path.isdir(target):
+            mkdir(target)
+        # sync files and directories to source not present in target
+        for (wd_source, dirs, files) in os.walk(source):
+            wd_target = os.path.normpath(
+                os.path.join(target, os.path.relpath(wd_source, source))
+            )
+            for dir in dirs:
+                mkdir(os.path.join(wd_target, dir))
+            for file in files:
+                file_source = os.path.join(wd_source, file)
+                file_target = os.path.join(wd_target, file)
+                if not os.path.exists(file_target):
+                    os.link(file_source, file_target)
+
+        # delete files and record directories from target not present in source
+        target_dirs = []
+        for (cwd, dirs, files) in os.walk(target):
+            wd_source = os.path.normpath(
+                os.path.join(source, os.path.relpath(cwd, target))
+            )
+
+            for file in files:
+                target_file = os.path.join(cwd, file)
+                source_file = os.path.join(wd_source, file)
+                if not os.path.isfile(source_file):
+                    rm(target_file)
+
+            for dir in dirs:
+                target_dir = os.path.join(cwd, dir)
+                source_dir = os.path.join(wd_source, dir)
+                if not os.path.isdir(source_dir):
+                    target_dirs.append(target_dir)
+
+        # delete directories from target not present in source
+        for dir in target_dirs:
+            rm(dir)
```

Comment:
I'm trying to understand why all this code is required. Is just calling `os.link(source, target)` insufficient?

Reply:
It is not needed for the initial run, but will most likely be necessary on sequential runs.

Say your source repo has three files, a, b & c. On the initial run you will get links to a, b & c in your target directory.

Now let's say the auther of the source repo adds file d and removes a, meaning files b, c & d are the content on the source. Without this extra code, you will however end up with a, b, c & d on the target.